### PR TITLE
fix(rome_js_analyze): check if the string key can convert to a static member

### LINKED ISSUE #4370: 🐛 `useLiteralKeys` doesn't ignore valid uses of square bracket notation
### Environment information

```block
CLI:
  Version:                      12.0.0-nightly.0133f40
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  ROME_LOG_DIR:                 unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v16.14.2"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/9.6.3"

Rome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           true
  Linter disabled:              false
  Organize imports disabled:    true

Workspace:
  Open Documents:               0
```

### What happened?

This may be as intended, but then it differs in an important way from eslint `dot-notation`.

See below

wat

There is no way to specify a literal key while also allowing special characters, the rule must then be ignored to allow this line to happen. Sometimes square bracket notation is unavoidable.

### Expected result

The rule should allow valid uses of square bracket notation

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

### PR DESCRIPTION
**Pull request by nissy-dev**
## Summary

Fix #4370

## Test Plan

I added some tests.

## Changelog

- [ ] The PR requires a changelog line

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs [hunk 1]
```diff
@@ -167,3 +171,14 @@ impl Rule for UseLiteralKeys {
...
     }
 }
+
+// check if the string key can convert to a static member
+fn is_convert_key_string_to_static_member(key: &str) -> bool {
+    key.chars().enumerate().all(|(index, c)| {
+        if index == 0 {
+            c.is_alphabetic() || c == '_'
+        } else {
+            c.is_alphabetic() || c.is_ascii_digit() || c == '_'
```
Comment (Conaclos) [inline crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs:181]: A JavaScript identifier can also contain `$` and some Unicode characters and Unicode escape sequences (this should be normalized by the formatter?). See [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#identifiers). I do not know if we can reuse some functions of the parser/lexer?

Comment (nissy-dev) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs:181]: Thank you! I have modified the logic to check for `$`.  

We can convert Unicode characters and Unicode escape sequences to static member, but I seem it is a little difficult to implement. And then, dot-notaion rule (similar ESLint rule) also doesn't check them. (see: https://github.com/eslint/eslint/blob/1fea2797801a82a2718814c83dad641dab092bcc/lib/rules/dot-notation.js#L18)


Therefore, I don't think we care about some Unicode characters and Unicode escape sequences. The main purpose of this PR reduces false positives.

Comment (nissy-dev) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: > I do not know if we can reuse some functions of the parser/lexer for checking if a string is a JS identifier?

I investigated, but couldn't find it. Does this function seem relevant...?
https://github.com/rome/tools/blob/7cfaf392e0f3edbbf37fde0fb237ee7b972d19be/crates/rome_js_parser/src/lexer/mod.rs#L836

Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: [`is_id_start` and `is_id_continue`](https://github.com/rome/tools/blob/main/crates/rome_js_unicode_table/src/lib.rs) sound interesting.

Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: > We can convert Unicode characters and Unicode escape sequences to static member, but I seem it is a little difficult to implement.

Could you be a bit more precise about what is challenging?

Comment (nissy-dev) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: I tried to add an test case like `a['\u4f60\u597d\'];`. 

If  key is `\u4f60\u597d`,  the result of `key.chars()` are `Chars(['\\', 'u', '4', 'f', '6', '0', '\\', 'u', '5', '9', '7', 'd'])`.  I don't comes up with any ideas of handling this. If you have some ideas, I would appreciate your advice.

File: crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs [hunk 2]
```diff
@@ -94,8 +94,10 @@ impl Rule for UseLiteralKeys {
             AnyJsExpression::AnyJsLiteralExpression(
                 AnyJsLiteralExpression::JsStringLiteralExpression(string_literal),
             ) => {
-                let value = string_literal.inner_string_text().ok()?;
-                return Some((string_literal.range(), value.to_string()));
+                let value = string_literal.inner_string_text().ok()?.to_string();
+                if is_convert_key_string_to_static_member(&value) {
+                    return Some((string_literal.range(), value));
+                }
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs:100]: nit: we can avoid calling the `to_string` at line 97, and only calling it inside the `if`. Doing so, we allocate a new `String` only when we need it

File: crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs [hunk 3]
```diff
@@ -167,3 +171,14 @@ impl Rule for UseLiteralKeys {
         }
     }
 }
+
+// check if the string key can convert to a static member
+fn is_convert_key_string_to_static_member(key: &str) -> bool {
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs:176]: nit: `can_convert_to_static_member` ?

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs (hunk 1 above)
Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: We could just rely on `is_id_start` and `is_id_continue` and leave out escaped unicode chars? A comment to say we do no tsupport escaped unicode chars could be a useful info. What do you think?

Comment (nissy-dev) [reply crates/rome_js_analyze/src/analyzers/nursery/use_literal_keys.rs]: I updated!

**Pull request merged**